File: 01_ML/01_supervised/01_regression/02_nonlinear_reg/nodejs/02d_exponential.js
// ML: Non-Linear Regression - Exponential
// (Node.js port of ../jupyter/02d_exponential.ipynb)
//
// Years held vs property value: compound appreciation at ~6 %/yr.
//
//   value = a * exp(r * years)
//
// Trick: fit a straight line on (x, ln(y)). Recover a = exp(intercept), r = slope.

import MLR from 'ml-regression-multivariate-linear';
import { writePlotlyHtml, linspace } from './plot.js';

const data = {
  years: [0, 2, 5, 8, 12, 15, 18, 22, 25, 30],
  value: [400, 450, 535, 640, 815, 970, 1160, 1465, 1720, 2300],
};
const test = { years: [3, 10, 20, 28] };

console.table(data.years.map((y, i) => ({ years: y, value: data.value[i] })));

const X = data.years.map((x) => [x]);
const Y = data.value.map((v) => [Math.log(v)]);

const reg = new MLR(X, Y);

const r = reg.weights[0][0];
const a = Math.exp(reg.weights[1][0]);
console.log(`fit: value = ${a.toFixed(4)} * exp(${r.toFixed(6)} * years)`);
console.log(`     implied annual appreciation: ${((Math.exp(r) - 1) * 100).toFixed(2)}%`);

const predictions = test.years.map((x) => a * Math.exp(r * x));
console.table(test.years.map((x, i) => ({ years: x, value: predictions[i] })));

const yearGrid = linspace(Math.min(...data.years), Math.max(...data.years), 200);
const curve = yearGrid.map((x) => a * Math.exp(r * x));

writePlotlyHtml(
  new URL('./02d_exponential_plot.html', import.meta.url).pathname,
  'Exponential Regression - years held vs value',
  [
    { type: 'scatter', mode: 'markers', name: 'training',
      x: data.years, y: data.value, marker: { color: 'magenta', symbol: 'star', size: 12 } },
    { type: 'scatter', mode: 'lines', name: 'exponential fit',
      x: yearGrid, y: curve, line: { color: 'blue' } },
    { type: 'scatter', mode: 'markers', name: 'predicted',
      x: test.years, y: predictions, marker: { color: 'red', size: 10 } },
  ],
  { xaxis: { title: 'years held' }, yaxis: { title: 'property value ($k)' } },
);

console.log('\nWrote 02d_exponential_plot.html');

File: 01_ML/01_supervised/01_regression/02_nonlinear_reg/python/02d_exponential.py
"""ML: Non-Linear Regression - Exponential.

Years held vs property value: a starter home appreciating at ~6 %/yr compounds
exponentially.

    value = a * exp(r * years)

Trick: take ln(y) of the target and fit a regular linear regression on (x, ln(y)).
Recover a = exp(intercept) and r = slope.

Run:
    python 02d_exponential.py
"""

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

PLOT_PATH = "02d_exponential_plot.png"


def main() -> None:
    data = {
        "years":  [0,   2,   5,   8,   12,  15,  18,  22,  25,  30],
        "value":  [400, 450, 535, 640, 815, 970, 1160, 1465, 1720, 2300],
    }
    test = {"years": [3, 10, 20, 28]}

    df = pd.DataFrame(data)
    print("training data (value in $k):")
    print(df.to_string(index=False))

    ln_value = np.log(df["value"])
    reg = LinearRegression().fit(df[["years"]].values, ln_value)

    r = reg.coef_[0]
    a = float(np.exp(reg.intercept_))
    print(f"\nfit: value = {a:.4f} * exp({r:.6f} * years)")
    print(f"     implied annual appreciation: {(np.exp(r) - 1) * 100:.2f}%")

    tst = pd.DataFrame(test)
    tst["value"] = a * np.exp(r * tst["years"])
    print("\npredictions:")
    print(tst.to_string(index=False))

    year_grid = np.linspace(df.years.min(), df.years.max(), 200)
    curve = a * np.exp(r * year_grid)

    plt.figure()
    plt.xlabel("years held")
    plt.ylabel("property value ($k)")
    plt.scatter(df.years, df.value, color="magenta", marker="*", s=80, label="training")
    plt.plot(year_grid, curve, color="blue", label="exponential fit")
    plt.scatter(tst.years, tst.value, color="red", label="predicted")
    plt.legend()
    plt.savefig(PLOT_PATH, bbox_inches="tight")
    print(f"\nWrote {PLOT_PATH}")


if __name__ == "__main__":
    main()
